### PR TITLE
Feature: add paused state

### PR DESCRIPTION
**Pull request by Axel-CH**
## Summary

The goal of this PR is to add a new state to the bot: **paused**

Follow up of the PR: [#11503](https://github.com/freqtrade/freqtrade/pull/11503)

## Quick changelog

- add paused state to initial state in schema
- add paused state to State class
- add paused state handler to rpc
- add paused state path to api
- allow entry only if state is running

## What's new?
With this new state the bot will be able to manage already open trades, including adjusting position, but won't be able to enter new positions.

The bot will be able to pass from any state to the other:

stopped <-> running <-> paused
paused <-> running <-> stopped
stopped <-> paused <-> running

### CODE REVIEW

File: docs/advanced-setup.md [hunk 1]
```diff
@@ -177,7 +177,7 @@ sudo loginctl enable-linger "$USER"
 If you run the bot as a service, you can use systemd service manager as a software watchdog monitoring freqtrade bot 
 state and restarting it in the case of failures. If the `internals.sd_notify` parameter is set to true in the 
 configuration or the `--sd-notify` command line option is used, the bot will send keep-alive ping messages to systemd 
-using the sd_notify (systemd notifications) protocol and will also tell systemd its current state (Running or Stopped) 
+using the sd_notify (systemd notifications) protocol and will also tell systemd its current state (Running or Paused or Stopped)
```
Comment (xmatthias) [inline docs/advanced-setup.md]:
```suggestion
using the sd_notify (systemd notifications) protocol and will also tell systemd its current state (Running, Paused or Stopped) 
```

same probably in other spots that were changed ... ;)

File: freqtrade/rpc/telegram.py [hunk 2]
```diff
@@ -170,14 +170,15 @@ def _init_keyboard(self) -> None:
         self._keyboard: list[list[str | KeyboardButton]] = [
             ["/daily", "/profit", "/balance"],
             ["/status", "/status table", "/performance"],
-            ["/count", "/start", "/stop", "/help"],
+            ["/start", "/pause", "/stop", "/help"],
```
Comment (xmatthias) [inline freqtrade/rpc/telegram.py]: i don't see any reason to change the default keyboard at this point.

**Review by xmatthias — CHANGES_REQUESTED** [review]
I don't see why we would need a new endpoint handling for this (essentially in all files within rpc).

Based on the discussion in the previous PR, i'd expect this to work exactly the same as `/stopbuy` (open trades are still processed, new ones are not).
What i'd expect is that behavior of `/stopbuy` is updated - eventually the endpoint is renamed / aliased with `/pause`.

However i clearly don't see a reason to have 2 completely different methods to accomplish the same thing - this will lead to confusion (people expect to restart their bot with `/start` - but stopped it with `/stopbuy` - so restarting won't work like this.

As an aside: i'd apreciate if you could **test** the new functionality (if it touches multiple RPC handlers, you'll have to manually test them all unfortunately). I can see that you never ran `/pause` through telegram - as it's not working currently.
While tests pass - the promised functionality isn't actually provided.

File: freqtrade/rpc/telegram.py (hunk 2 above)
Comment (Axel-CH) [reply freqtrade/rpc/telegram.py]: I think it could be relevant to add the **pause** button in the third row with **start** and **stop**. Following this idea, count can be replaced.
I don't know how often the count button is used but as it does not provide much information can we combine the output with status table? Of course this could be done in another PR.

Comment (xmatthias) [reply freqtrade/rpc/telegram.py]: if that's the argument, then the whole 3rd row (except for help) is pointless, as i know near no-one that uses `/start` and `/stop` on a regular basis.

Comment (Axel-CH) [reply freqtrade/rpc/telegram.py]: Well as this is a customizable table, users can change that if they want, so I'll not try to guess what they need.

It just seem logic to me to have pause with start and stop, as on a media player or a tv remote 😅

Comment (xmatthias) [reply freqtrade/rpc/telegram.py]: i'd leave it as is for now - this an be changed at a later point, IF we see that people actually use this :)

Comment (Axel-CH) [reply freqtrade/rpc/telegram.py]: Let's do this, so I revert that 👍🏽

**Comment by Axel-CH** [comment]
I made the latest requested changes by combining related endpoints.
Performed manual testing on state changes and fixed the `/pause` Telegram command.

I’ll also add a `test_worker_lifecycle` test to cover the new possible transitions between states.
Aside from that, I think we’re good.

File: freqtrade/worker.py [hunk 3]
```diff
@@ -96,7 +96,7 @@ def _worker(self, old_state: State | None) -> State:
             logger.info(
                 f"Changing state{f' from {old_state.name}' if old_state else ''} to: {state.name}"
             )
-            if state == State.RUNNING:
+            if old_state == State.STOPPED and (state == State.RUNNING or state == State.PAUSED):
```
Comment (xmatthias) [inline freqtrade/worker.py]: I'm not sure this is correct (though i didn't have time to fully test this in detail) - ignoring the fact that i don't like "or" if it could be simpified by `in` for now :laughing: 

Previous behavior was to call startup if the new state is running - which is guarded by the parent if of `old_state != state`.

So  `.startup()` is called e.g. if the bot changes from reload_config to "running".
This ain't happening now - so i think this line will break proper `/reload_config` workings - causing the bot to remain in a "stalled" position.

It also seems to prevent telegram startup messages completely (on a freshly started bot) - so this seems to introduce a bug.

File: freqtrade/worker.py [hunk 4]
```diff
@@ -112,9 +112,10 @@ def _worker(self, old_state: State | None) -> State:
 
             self._throttle(func=self._process_stopped, throttle_secs=self._throttle_secs)
 
-        elif state == State.RUNNING:
+        elif state == State.RUNNING or state == State.PAUSED:
```
Comment (xmatthias) [inline freqtrade/worker.py]:
```suggestion
        elif state in (State.RUNNING, State.PAUSED):
```

File: freqtrade/rpc/rpc.py [hunk 5]
```diff
@@ -838,7 +838,7 @@ def _rpc_start(self) -> dict[str, str]:
 
     def _rpc_stop(self) -> dict[str, str]:
         """Handler for stop"""
-        if self._freqtrade.state == State.RUNNING:
+        if self._freqtrade.state == State.RUNNING or self._freqtrade.state == State.PAUSED:
```
Comment (xmatthias) [inline freqtrade/rpc/rpc.py]:
```suggestion
        if self._freqtrade.state != State.STOPPED:
```

**Review by xmatthias — COMMENTED** [review]
Aside from the introduced bug (missing messages, ...) - theres one other thing that jumped to my mind (from a behavior perspective).

while "/pause" will prevent the bot from entering new trades - it won't stop an open position from increasing.
current suggestion was to check `self.max_open_trades` in the callback - which won't work any longer.
So i think we have to improve behavior by having "/pause" only allow position decreases (or exits) - but no new entries or position increases.

**Comment by Axel-CH** [comment]
Your suggestion to only allow position size decreases in the paused state actually makes sense. I'll introduce that tomorrow.

**Comment by xmatthias** [comment]
> Your suggestion to only allow position size decreases in the paused state actually makes sense. I'll introduce that tomorrow.

It's been a (known) limitation of the previous "paused" state - one that wasn't possible to circumvent without introducing a new state :laughing:

**Comment by Axel-CH** [comment]
When we set the bot in paused state, what do you think we should do with existing on exchange entry order from previous position adjustment call?
We keep or we cancel them?

I would keep them as they where created before the change to paused state.

**Comment by xmatthias** [comment]
keeping is fine, yes.
it's not a "new" entry - just handling of what was previously created.

File: freqtrade/worker.py (hunk 3 above)
Comment (Axel-CH) [reply freqtrade/worker.py]: I changed that and did some manual testing changing the state from config file using the /reload_config command. It's working now 👍🏽

File: tests/freqtradebot/test_worker.py [hunk 6]
```diff
@@ -50,6 +70,73 @@ def test_worker_stopped(mocker, default_conf, caplog) -> None:
...
+            False,
+            "Changing state from PAUSED to: STOPPED",
+        ),
+        (
+            State.RELOAD_CONFIG,
+            State.RUNNING,
+            State.RUNNING,
+            True,
+            "Changing state from RELOAD_CONFIG to: RUNNING",
+        ),
```
Comment (xmatthias) [inline tests/freqtradebot/test_worker.py:104]: I've modified this test slightly so it shows if "startup" was called or not.

i think this test (paused -> running) shows an edge-case - would we expect "startup" to be called here?

File: tests/freqtradebot/test_worker.py [hunk 7]
```diff
@@ -50,6 +70,54 @@ def test_worker_stopped(mocker, default_conf, caplog) -> None:
     assert mock_throttle.call_count == 1
 
 
+@pytest.mark.parametrize(
+    "old_state,target_state,startup_call,log_fragment",
+    [
+        (State.STOPPED, State.PAUSED, True, "Changing state from STOPPED to: PAUSED"),
+        (State.RUNNING, State.PAUSED, False, "Changing state from RUNNING to: PAUSED"),
+        (State.PAUSED, State.RUNNING, False, "Changing state from PAUSED to: RUNNING"),
```
Comment (xmatthias) [inline tests/freqtradebot/test_worker.py:78]: I've modified this test slightly so it shows if "startup" was called or not.

i think this test (paused -> running) shows an edge-case - would we expect "startup" to be called here?
(for clarity: i'm not saying "change this" - i'm more asking if this is what we'd expect here)

Comment (Axel-CH) [reply tests/freqtradebot/test_worker.py:78]: Indeed it's an edge case, I don't think it's relevant to call `startup` function as the `paused` state is kind of a `running` state without entry.
So there is no point to call the majority of functions that are within `startup`

**Review by Axel-CH — COMMENTED** [review]
Looks good to me, do you think we need to add or change something?

File: freqtrade/rpc/rpc.py [hunk 8]
```diff
@@ -849,16 +849,19 @@ def _rpc_reload_config(self) -> dict[str, str]:
...
         """
-        Handler to stop buying, but handle open trades gracefully.
+        Handler to pause trading (stop entering new trades), but handle open trades gracefully.
         """
         if self._freqtrade.state == State.RUNNING:
-            # Set 'max_open_trades' to 0
-            self._freqtrade.config["max_open_trades"] = 0
-            self._freqtrade.strategy.max_open_trades = 0
+            self._freqtrade.state = State.PAUSED
+            return {"status": "pausing trader ..."}
```
Comment (xmatthias) [inline freqtrade/rpc/rpc.py]: I'm not sure i like this "return" here.
i understand the intend - but it hides the essential info (how can you start it again?) - which deviates from prior behavior, where the instruction was included right there.

it's a direct message that's returned - but wording of the fallthrough (L864) would be clearer as to what happened - so maybe we should remove this line in favor of that. (L862 is fine i think - as that's a different state change).

image

Comment (Axel-CH) [reply freqtrade/rpc/rpc.py]: I've added the **Run /start to enable entries**. I also created a small wording difference when the bot just change to paused state or if it is already paused and the user trigger again `/pause`

File: freqtrade/rpc/rpc.py [hunk 9]
```diff
@@ -849,16 +849,26 @@ def _rpc_reload_config(self) -> dict[str, str]:
...
+        Handler to pause trading (stop entering new trades), but handle open trades gracefully.
         """
         if self._freqtrade.state == State.RUNNING:
-            # Set 'max_open_trades' to 0
-            self._freqtrade.config["max_open_trades"] = 0
-            self._freqtrade.strategy.max_open_trades = 0
+            self._freqtrade.state = State.PAUSED
+
+        if self._freqtrade.state == State.PAUSED:
+            return {"status": "paused, no entries will occur. Run /start to enable entries."}
```
Comment (xmatthias) [inline freqtrade/rpc/rpc.py]: why can't we let this "fall through" to the last line? 

i understand you _tried_ to introduce a wording difference - but it doesn't work that way - as you set the state above - so if the state was paused **or running**, you'll never fall through to the last line.
if you're already paused - this line will also trigger.

Comment (Axel-CH) [reply freqtrade/rpc/rpc.py]: I did the changes, It's ok now

File: freqtrade/rpc/rpc.py [hunk 10]
```diff
@@ -849,16 +849,23 @@ def _rpc_reload_config(self) -> dict[str, str]:
...
-            # Set 'max_open_trades' to 0
-            self._freqtrade.config["max_open_trades"] = 0
-            self._freqtrade.strategy.max_open_trades = 0
+            self._freqtrade.state = State.PAUSED
 
-        return {"status": "No more entries will occur from now. Run /reload_config to reset."}
+        if self._freqtrade.state == State.STOPPED:
+            self._freqtrade.state = State.PAUSED
+            return {
+                "status": "starting bot with trader in paused state, no entries will occur. \
```
Comment (xmatthias) [inline freqtrade/rpc/rpc.py]: Don't split strings like this - that's bad for 2 reasons
1 - splitting lines with `\` is very error prone
2 - it causes rendering issues

image

**Review by xmatthias — APPROVED** [review]
LGTM :+1:
